### DISCUSSION TITLE
Is it possible to de-dupe the stringification of queries?

### DISCUSSION THREAD

**Discussion by nathanmarks** [urql-graphql/urql#2714] [Q&A] [answered]
When a lot of components mount at once pointing at a hook that has the same `useQuery`, there's a decent amount of time spent stringifiying the query for every `useQuery` instance for the same query.

Even though the persisted query exchange gives us the ability to inject our own hash function that can look at a cache / use pre-computed shas in the build, we can't avoid the serialisation that happens here in `makeFetchBody`: https://github.com/FormidableLabs/urql/blob/b9ca5d3e39565bdc5dff61bd66e6e629ec3618ce/exchanges/persisted-fetch/src/persistedFetchExchange.ts#L63

Ultimately, this string gets thrown away when using persisted queries, so it's frustrating that we can't skip it in any way if we don't need it and can just hash once without the pre-stringified query in our custom hash fn and memoize that so we don't incur the cost every time.

But even if we were using the regular fetch exchange, it would be nice to de-dupe this. Is there a reason this isn't deduped today or something in the architecture of urql that makes it problematic to do that? 

If we could make `makeFetchBody` an option for the exchanges, it would mean we can achieve this without having to reimplement the exchange ourselves.

**Comment by JoviDeCroock** [comment, marked as answer]
The `dedupExchange` should ensure you only reach the `fetchExchange` once for a query (even if it's mounted by multiple `useQuery` invocations), alternatively you can always leverage your own `fetchExchange` we export the most needed parts from core.

**Reply by nathanmarks** [reply]
@JoviDeCroock i'm using dedup exchange, and the API calls themselves are deduped, but every single useQuery is somehow making it to the `makeFetchBody` line

**Reply by JoviDeCroock** [reply]
That could be the synchronous cache-check, anyway the latter point still stands 😅

**Reply by nathanmarks** [reply]
Haha yeah, that's what i've done for now (forked it).

If it's a synchronous cache-check, why does it make it all the way to fetch exchange? I haven't spent much time understanding the middleware architecture.

**Reply by JoviDeCroock** [reply]
The fetchSource itself doesn't trigger but the makeFetchBody etc happen outside of that which will happen. The source is deferred by a microtick so it cancels itself after a sync check

**Reply by nathanmarks** [reply]
Is there a reasonable/obvious path to optimising this? Or making `makeFetchBody` configurable on the exchange? Or is the only option to maintain a fork of the exchange.
